snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/request/SnomedQueryOptimizer.java

```diff
@@ -15,169 +15,847 @@
+		hierarchyStats.filterByClusterSize(minimumClusterSize);
+		// b) a non-member with a single member as its child -- there is no reason to replace "=child" with "<parent" in this case
+		hierarchyStats.filterSingleChildMember();
+		// c) more false positive children than true positive children
+		hierarchyStats.filterByChildren();
+		// d) ancestor has no false positives, but it has an _ancestor_ which has no false positives either (or at most 1 and is a non-member)
+		hierarchyStats.filterRedundantNoFalsePositives(conceptsToExclude);
+	}	
+
+	// Compute scores, then iterate over the best candidates with an acceptable fit in decreasing order
+	private void optimizeAncestorInclusions(final BranchContext context, final SnomedHierarchyStats inclusionHierarchyStats) {
+
+		final LongKeyFloatMap ancestorScores = inclusionHierarchyStats.initAncestorScores(
+			context, 
+			conceptSet, 
+			optimizerStrategy, 
+			zoom, 
+			weight);
+
+		if (ancestorScores.isEmpty()) {
+			return;
+		}
+		
+		final Instant startTime = Instant.now(clock);
+		final Instant endTime = startTime.plus(maxRuntime);
+
+		int iteration = 1;
+		int ancestorExpressionCount = 0;
+		boolean ancestorFound = false;
+		boolean canceled = false;
+
+		while (!canceled) {
+			final String bestFitAncestor = getBestAncestor(context, ancestorScores);
+
+			if (bestFitAncestor != null) {
+				ancestorFound = true;
+				ancestorExpressionCount++;
+
+				// Convert ancestor to an inclusion expression
+				final boolean member = conceptSet.contains(bestFitAncestor);
+				final String operator = member ? "<<" : "<";
+				final QueryExpression bestFitInclusion = new QueryExpression(IDs.base62UUID(), String.format("%s %s", operator, bestFitAncestor), false);
+
+				// Add inclusion to our optimized list
+				optimizedInclusions.add(bestFitInclusion);
+
+				// "applyInclusions" is effectively inlined here because we need access to the number of concepts included
+				final Set<String> includedIds = evaluateEclToSet(context, bestFitInclusion.getQuery());
+				final Set<String> expected = Sets.intersection(includedIds, conceptSet);
+				conceptsToInclude.removeAll(expected);
+
+				final Set<String> unexpected = Sets.difference(includedIds, conceptSet);
+				conceptsToExclude.addAll(unexpected);
+
+				// Tune parameters after a successful selection
+				if (((float) conceptSet.size()) / (includedIds.size() + 1) > 1000.0f) {
+					final float newFitThreshold = fitThreshold * 0.9f;
+					if (newFitThreshold > minimumFitThreshold) {
+						fitThreshold = newFitThreshold;
+						zoom = idealClauseCount;
+						log.info("Fit threshold changed to {}, zoom is {}", fitThreshold, zoom);
+					}
+				}
+
+				// Elevate strategy after using the same one for 100 iterations (but don't promote to LOSSY from non-lossy strategies)
+				if (iteration > (optimizerStrategy.ordinal() + 1) * 100 && optimizerStrategy.ordinal() < OptimizerStrategy.values().length - 2) {
```

Comment:
Ohmy, enum ordinal magic, would it be possible to refactor this to something readable? 

Something like this for example:
```
if (optimizerStrategy.isElevateNeeded(iterations)) {
  optimizerStrategy = optimizerStrategy.elevate();
}
```
And document elevation strategy in the enum type.
Thanks!